### PR TITLE
[libp2p-dns] Use trust-dns-resolver (with either async-std or tokio). Remove thread pool.

### PR DESCRIPTION
**Pull request by romanb**
This PR changes `libp2p-dns` to use the `trust-dns-resolver` library for DNS resolution, thereby removing the current use of the thread pool, i.e. closing https://github.com/libp2p/rust-libp2p/issues/1235. This also serves as a first step for https://github.com/libp2p/rust-libp2p/issues/1920, which will build on this work. 

Since `trust-dns-resolver` and related crates already provide decent support for both `async-std` and `tokio`, we make use of that here in `libp2p-dns`, offering analogous `async-std` and `tokio` features. Just like with `libp2p-tcp`, the `async-std` variant is enabled by default. This PR is only pending a new release that includes https://github.com/bluejekyll/trust-dns/pull/1354, which is why it is still pointing to a git dependency. Also just like with `libp2p-tcp`, the `libp2p-dns` tests are exercised for both runtimes.

Since `trust-dns-resolver` provides many useful configuration options and error details, central types of `trust-dns-resolver` like `ResolverConfig`, `ResolverOpts` and `ResolveError` are re-exposed in the API of `libp2p-dns`. Full encapsulation
does not seem preferable in this case. We also re-export the (currently tokio-only) features for DNS-over-TLS and DNS-over-HTTPS (via rustls).

I also consolidated the top-level utility functions of the `libp2p` crate: There is now just `fn development_transport()` (available with default features) and `fn tokio_development_transport()` (available when the corresponding tokio features are enabled). I think that is sufficient and I removed the minor variation that also included `pnet` support into the development transport.

### CODE REVIEW

File: transports/dns/src/lib.rs [hunk 1]
```diff
@@ -299,28 +351,69 @@ mod tests {
...
                 .dial("/ip4/1.2.3.4/tcp/20000".parse().unwrap())
                 .unwrap()
                 .await
                 .unwrap();
-        });
+
+            // Failure due to no records.
+            match transport
+                .clone()
+                .dial("/dns4/domaindoesnotexistforsureamiright.com/tcp/20000".parse().unwrap())
```
Comment (mxinden) [inline transports/dns/src/lib.rs:389]: :slightly_smiling_face: If I understand correctly an alternative would be [`invalid.`](https://en.wikipedia.org/wiki/.invalid) defined by the IETF.

**Review by mxinden — APPROVED** [review]
Code changes look good to me.

I don't hold a strong opinion on OS resolver vs. library resolver. In general I would consider DNS resolution a responsibility of the OS, at the same time https://github.com/bluejekyll/trust-dns seems to be well established and thus worth betting on.

File: transports/dns/Cargo.toml [hunk 2]
```diff
@@ -13,3 +13,21 @@ categories = ["network-programming", "asynchronous"]
 libp2p-core = { version = "0.27.0", path = "../../core" }
 log = "0.4.1"
 futures = "0.3.1"
+trust-dns-resolver = { version = "0.20", default-features = false, features = ["system-config"], git = "https://github.com/bluejekyll/trust-dns", branch = "main" }
```
[the remark this answers is not in the document]
Comment (romanb) [reply transports/dns/Cargo.toml:16]: Yes, but (quoting from the PR description):

> This PR is only pending a new release that includes bluejekyll/trust-dns#1354, which is why it is still pointing to a git dependency.

[the remark this answers is not in the document]
Comment (mxinden) [reply transports/dns/Cargo.toml:16]: :+1:  Oh, I missed that. Sorry.

File: transports/dns/src/lib.rs (hunk 1 above)
Comment (romanb) [reply transports/dns/src/lib.rs]: Done in https://github.com/libp2p/rust-libp2p/pull/1927/commits/37a02b9dc70e9732a4b7e6887929a0b781dd7c31.

File: transports/dns/src/lib.rs [hunk 3]
```diff
@@ -20,85 +20,135 @@
...
-/// Represents the configuration for a DNS transport capability of libp2p.
-///
-/// This struct implements the `Transport` trait and holds an underlying transport. Any call to
-/// `dial` with a multiaddr that contains `/dns/`, `/dns4/`, or `/dns6/` will be first be resolved,
-/// then passed to the underlying transport.
-///
-/// Listening is unaffected.
+use log::{debug, trace};
+use std::{error, fmt, net::IpAddr};
+#[allow(unused_imports)]
```
[the remark this answers is not in the document]
Comment (romanb) [reply transports/dns/src/lib.rs]: Done in https://github.com/libp2p/rust-libp2p/pull/1927/commits/37a02b9dc70e9732a4b7e6887929a0b781dd7c31.

**Comment by romanb** [comment]
Yes, of course. I am anyway already working on the follow-up PR for `/dnsaddr`. Regarding your comment of OS resolver vs library resolver, apart from more features and better configurability from the code, a main motivation for introducing `trust-dns-resolver` here is that it makes it easy to do the necessary TXT lookups for `/dnsaddr`. I also think that these DNS lookups via `trust-dns-resolver` will actually perform better as they're more targeted to specific record types, as opposed to the opaque use of [`ToSocketAddrs`](https://doc.rust-lang.org/std/net/trait.ToSocketAddrs.html) from a background thread.

**Comment by romanb** [comment]
Unfortunately still pending a new release of `async-std-resolver`.

**Comment by mxinden** [comment]
> Unfortunately still pending a new release of `async-std-resolver`.

@romanb are you aware of a timeline for this release to happen? I am happy to reach out to the authors myself. It is not urgent, though it would be nice to build on top of the general changes introduced via https://github.com/libp2p/rust-libp2p/pull/1931. In addition it would be a bummer for this pull request and current `master` to drift apart.

**Comment by romanb** [comment]
> are you aware of a timeline for this release to happen?

No, but based on the prior release history of the crate I was indeed hoping for a new release some time in February. It would be nice if you could reach out.

**Comment by romanb** [comment]
`trust-dns-0.20.1` has been released, so we can go ahead. Thanks for following up on it @mxinden.

**Pull request merged**